File: output_20260320_065752/claude-opus/c/Struct/deposit.c
#include "stdlib.h"

/*@ predicate account(struct account *a; int balance) =
      a->balance |-> balance;
@*/

struct account
{
    int balance;
};

/*@ 
  requires true;
  ensures account(result, 0);
@*/
struct account *create_account()
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    //@ close account(myAccount, 0);
    return myAccount;
}

/*@
  requires account(myAccount, ?balance) &*& 0 <= amount;
  ensures account(myAccount, balance + amount);
@*/
void account_deposit(struct account *myAccount, int amount)
{
    //@ open account(myAccount, ?balance);
    myAccount->balance += amount;
    //@ close account(myAccount, balance + amount);
}

/*@
  requires account(myAccount, ?balance);
  ensures account(myAccount, balance) &*& result == balance;
@*/
int account_get_balance(struct account *myAccount)
{
    //@ open account(myAccount, ?balance);
    int result = myAccount->balance;
    //@ close account(myAccount, balance);
    return result;
}

/*@
  requires account(myAccount, ?balance);
  ensures account(myAccount, newBalance);
@*/
void account_set_balance(struct account *myAccount, int newBalance)
{
    //@ open account(myAccount, _);
    myAccount->balance = newBalance;
    //@ close account(myAccount, newBalance);
}

/*@
  requires account(myAccount, _);
  ensures true;
@*/
void account_dispose(struct account *myAccount)
{
    //@ open account(myAccount, _);
    free(myAccount);
}

int main()
/*@
  requires true;
  ensures true;
@*/
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output_log/claude-opus/sample_5/rust/qwen3/rust/purely_unsafe/solutions/deposit.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

#[repr(C)]
pub struct Account {
    balance: i32,
}

#[predicate]
pub account_pred(p: *mut Account) =
    p != 0 &*& malloc_block_Account(p, 1) &*& acc_balance_field(p, ?b) &*& *p |-> Account { balance: b };

#[predicate]
pub acc_balance_field(p: *mut Account, b: i32) =
    p != 0 &*& *p |-> Account { balance: b };

#[lemma]
fn account_balance_access(p: *mut Account)
    requires account_pred(p)
    ensures account_pred(p) &*& exists(?b) &*& acc_balance_field(p, b)
{
    open account_pred(p);
    close account_pred(p);
}

impl Account {
    unsafe fn create() -> *mut Account {
        #[requires(true)]
        #[ensures(account_pred(result))]
        {
            let my_account = alloc(Layout::new::<Account>()) as *mut Account;
            if my_account.is_null() {
                handle_alloc_error(Layout::new::<Account>());
            }
            #[invariant(malloc_block_Account(my_account, 1))]
            {
                *my_account = Account { balance: 0 };
            }
            close account_pred(my_account);
            my_account
        }
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32 {
        #[requires(account_pred(my_account))]
        #[ensures(account_pred(my_account) &*& result == (*my_account).balance)]
        {
            open account_pred(my_account);
            let b = (*my_account).balance;
            close account_pred(my_account);
            b
        }
    }

    unsafe fn set_balance(my_account: *mut Account, new_balance: i32) {
        #[requires(account_pred(my_account))]
        #[ensures(account_pred(my_account) &*& (*my_account).balance == new_balance)]
        {
            open account_pred(my_account);
            *my_account = Account { balance: new_balance };
            close account_pred(my_account);
        }
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32) {
        #[requires(account_pred(my_account))]
        #[ensures(account_pred(my_account) &*& (*my_account).balance == old((*my_account).balance) + amount)]
        {
            open account_pred(my_account);
            let old_balance = (*my_account).balance;
            *my_account = Account { balance: old_balance + amount };
            close account_pred(my_account);
        }
    }

    unsafe fn dispose(my_account: *mut Account) {
        #[requires(account_pred(my_account))]
        #[ensures(true)]
        {
            open account_pred(my_account);
            dealloc(my_account as *mut u8, Layout::new::<Account>());
        }
    }
}

fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}